polkadot/node/core/candidate-validation/src/lib.rs

```diff
@@ -378,25 +552,26 @@ where
 			},
 		};
 
-	let executor_params =
-		if let Ok(executor_params) = executor_params_at_relay_parent(relay_parent, sender).await {
-			gum::debug!(
-				target: LOG_TARGET,
-				?relay_parent,
-				?validation_code_hash,
-				"precheck: acquired executor params for the session: {:?}",
-				executor_params,
-			);
-			executor_params
-		} else {
-			gum::warn!(
-				target: LOG_TARGET,
-				?relay_parent,
-				?validation_code_hash,
-				"precheck: failed to acquire executor params for the session, thus voting against.",
-			);
-			return PreCheckOutcome::Invalid
-		};
+	let executor_params = if let Ok(executor_params) =
```

Comment:
That's a good question indeed: What set of executor parameters should be used to compile the PVFs in advance? There may already be something new in `pending_config` that will become effective in the next session, and all this in-advance compilation will be in vain in that case. That's a corner case, as changes to the executor parameters are extremely rare, so I do not insist it should be addressed in this very PR, but it's something that might probably be considered in the future. Generally, it's not the first time we're having hard times trying to predict next session's executor parameters: #694

Reply:
It might not be so bad to precompile pvf anyway. Anyway, the validator will have to compile a new one in the next session. Or we can consider a smarter solution?

Reply:
I'm not 100% sure I'm not missing something, but it seems like we could use `pending_config` from the `configuration` pallet to get the executor environment params. That gives us two benefits:
1. We're always getting params that will be effective in the future session;
2. We get it directly from storage, thus saving a couple of runtime API calls.

@sandreim WDYT?

Reply:
Yes, we already know that in session N we are gonna become a parachain validator. So, we need to prepare the artifacts using the same execution params from session N.

I think 1 is preferable as a proper interaface, rather than reading storage directly.

Alternatively, we could could introduce a new runtime API that returns the execution parameters for a given session. This should solve future problems with getting the next session executor params in one place.

Reply:
@sandreim @s0me0ne-unkn0wn I think it can be a good job for an additional PR. What do you think if I move this conversation to an issue and start doing that after we merged this one?

Reply:
After recalling how it works, I agree. That will require changing the runtime API, putting that into the staging API, then bumping its version, etc. Not an easy change. Raised #5080.

Reply:
After a conversation with @sandreim:
- This concern is valid and not only for just connected validators but for all. See https://github.com/paritytech/polkadot-sdk/issues/4203
- It's OK to solve this issue separately. 

@s0me0ne-unkn0wn what do you think?

Reply:
I'm ok with implementing https://github.com/paritytech/polkadot-sdk/issues/4203 in a separate PR which applies to these changes as well.  

Getting this one in the current release if we can still make it to backport there would be great. I'd want to wait for validators to upgrade to this fix before bumping Polkadot validator count to 400.